File: laba4/laba4.go
package main

import (
	"fmt"
	"math"
)

// f(x) = x * ln(x+1) - 1
func f(x float64) float64 {
	return x*math.Log(x+1) - 1.0
}

// f'(x) = ln(x+1) + x/(x+1)
func df(x float64) float64 {
	return math.Log(x+1) + x/(x+1)
}

// φ(x) = 1/ln(x+1) для метода простых итераций
func phi(x float64) float64 {
	return 1.0 / math.Log(x+1)
}

// Структура для хранения данных итерации
type Iteration struct {
	Num  int     
	Xn   float64
	Xn1  float64 
	Diff float64 
}

// Структура для хранения интервала (метод половинного деления)
type Interval struct {
	A, B float64 
}

func printIterHeader(method string) {
	fmt.Printf("\n=== %s ===\n", method)
	fmt.Println("n\txn\t\t\txn+1\t\t\t|xn+1 - xn|")
}

func printIteration(it Iteration) {
	fmt.Printf("%d\t%.6f\t\t%.6f\t\t%.6f\n", it.Num, it.Xn, it.Xn1, it.Diff)
}

func bisection(a, b, eps float64) {
	fmt.Println("\n=== Метод половинного деления ===")
	fmt.Println("an\t\t\tbn\t\t\t(bn-an)")

	intervals := make([]Interval, 0)

	for (b - a) > eps {
		intervals = append(intervals, Interval{a, b})
		c := (a + b) / 2
		if f(a)*f(c) < 0 {
			b = c
		} else {
			a = c
		}
	}

	for _, interval := range intervals {
		fmt.Printf("%.6f\t\t%.6f\t\t%.6f\n", interval.A, interval.B, interval.B-interval.A)
	}

	root := (a + b) / 2
	fmt.Printf("\nКорень: %.6f\n", root)
	fmt.Printf("Итераций: %d\n", len(intervals))
}

func newton(x0, eps float64) {
	printIterHeader("Метод Ньютона")

	iterations := make([]Iteration, 0)
	n := 0
	x := x0

	for {
		xNext := x - f(x)/df(x) // формула Ньютона
		diff := math.Abs(xNext - x)

		iterations = append(iterations, Iteration{n, x, xNext, diff})

		if diff < eps {
			break
		}
		x = xNext
		n++
	}

	for _, it := range iterations {
		printIteration(it)
	}

	root := iterations[len(iterations)-1].Xn1
	fmt.Printf("\nКорень: %.6f\n", root)
	fmt.Printf("Итераций: %d\n", len(iterations))
}

func fixedPoint(x0, eps float64) {
	printIterHeader("Метод простых итераций")

	iterations := make([]Iteration, 0)
	n := 0
	x := x0

	for {
		xNext := phi(x) // x_{n+1} = φ(x_n)
		diff := math.Abs(xNext - x)

		iterations = append(iterations, Iteration{n, x, xNext, diff})

		if diff < eps {
			break
		}
		x = xNext
		n++
	}

	for _, it := range iterations {
		printIteration(it)
	}

	root := iterations[len(iterations)-1].Xn1
	fmt.Printf("\nКорень: %.6f\n", root)
	fmt.Printf("Итераций: %d\n", len(iterations))
}

func main() {
	// Отделение корней: f(1) < 0, f(2) > 0 → корень на [1, 2]
	fmt.Println("Уравнение: x * ln(x+1) - 1 = 0")
	fmt.Println("Интервал: [1.0, 2.0]")
	fmt.Println("Точность: 1e-4\n")

	// Запуск трёх методов
	bisection(1.0, 2.0, 1e-4)
	newton(1.5, 1e-4)
	fixedPoint(1.5, 1e-4)
}

File: laba4/laba4.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <iomanip>

using namespace std;

// Уравнение: x * ln(x+1) - 1 = 0 
auto f = [](double x) { return x * log(x + 1) - 1.0; };
auto df = [](double x) { return log(x + 1) + x / (x + 1); };
auto phi = [](double x) { return 1.0 / log(x + 1); }; // x = 1/ln(x+1)

void print_table_header(const string& method) {
    cout << "\n=== " << method << " ===\n";
    cout << "n\txn\t\txn+1\t\t|xn+1 - xn|\n";
}

void bisection(double a, double b, double eps) {
    cout << "\n=== Метод половинного деления ===\n";
    cout << "an\t\tbn\t\t(bn-an)\n";
    
    vector<pair<double, double>> intervals;
    while ((b - a) > eps) {
        intervals.push_back({a, b});
        double c = (a + b) / 2;
        (f(a) * f(c) < 0) ? b = c : a = c;
    }
    
    for (auto [l, r] : intervals) 
        cout << fixed << setprecision(6) << l << "\t" << r << "\t" << (r - l) << "\n";
    
    cout << "\nКорень: " << (a + b) / 2 << "\nИтераций: " << intervals.size() << "\n";
}

void newton(double x0, double eps) {
    print_table_header("Метод Ньютона");
    
    vector<tuple<int, double, double, double>> iter;
    int n = 0;
    double x = x0;
    
    while (true) {
        double x_next = x - f(x) / df(x);
        iter.push_back({n, x, x_next, fabs(x_next - x)});
        if (fabs(x_next - x) < eps) break;
        x = x_next;
        n++;
    }
    
    for (auto [i, xn, xnp1, diff] : iter)
        cout << i << "\t" << fixed << setprecision(6) << xn << "\t" << xnp1 << "\t" << diff << "\n";
    
    cout << "\nКорень: " << get<2>(iter.back()) << "\nИтераций: " << iter.size() << "\n";
}

void fixed_point(double x0, double eps) {
    print_table_header("Метод простых итераций");
    
    vector<tuple<int, double, double, double>> iter;
    int n = 0;
    double x = x0;
    
    while (true) {
        double x_next = phi(x);
        iter.push_back({n, x, x_next, fabs(x_next - x)});
        if (fabs(x_next - x) < eps) break;
        x = x_next;
        n++;
    }
    
    for (auto [i, xn, xnp1, diff] : iter)
        cout << i << "\t" << fixed << setprecision(6) << xn << "\t" << xnp1 << "\t" << diff << "\n";
    
    cout << "\nКорень: " << get<2>(iter.back()) << "\nИтераций: " << iter.size() << "\n";
}

int main() {
    cout << "Уравнение: x * ln(x+1) - 1 = 0\nИнтервал: [1, 2]\nТочность: 1e-4\n";
    
    bisection(1.0, 2.0, 1e-4);
    newton(1.5, 1e-4);
    fixed_point(1.5, 1e-4);
    
    return 0;
}
